src/codecs/bmp/decoder.rs

```diff
@@ -1526,7 +1526,7 @@ impl<'a, R: 'a + Read + Seek> ImageDecoder<'a> for BmpDecoder<R> {
 }
 
 impl<'a, R: 'a + Read + Seek> ImageDecoderRect<'a> for BmpDecoder<R> {
-    fn read_rect_with_progress<F: Fn(Progress)>(
+    fn read_rect_with_progress<F: FnMut(Progress)>(
```

Comment:
We can also offer compatibility shims, since the callback is not a guarantee

```rust
let fn_progress = RefCell::new(progress);
read_rect_with_progress(self.0, buf, width, height, color_type, |progress| {
    if let Ok(fn_progress) = fn_progress.try_borrow_mut() {
        fn_progress(progress)
    }
});
```

Reply:
You mean as an additional helper that could be called by the broken downstream implementations to fix the progress? Yes, this could be a nice help for those implementations.

Reply:
Or as a backwards compatible implementation. Since `FnMut` is a super-trait of `Fn` it is certainly more generic for the caller, which would be encouraged for any new cases of `_with_progress` functions (#1695) in particular. But yet I'm not convinced that this warrants making a breaking change here. There's no logic bug or panicking in the above implementation—indeed the same could be written by any caller. This means that the generalization is merely a convenience, and breaking change for convenience isn't worth it.

So I would propose rather than _changing_ the existing cases we 
* add a new function `with_progress_mut`
* _maybe_ deprecate the existing function, although not immediately
* write it down for `0.25`  where we'll align the interfaces
* for _new_ interfaces we use `with_progress(impl FnMut)`

Reply:
Is there a branch for 0.25?

Reply:
Not yet, but we could make one